test/parallel_api/ranges/std_ranges_test.h

```diff
@@ -0,0 +1,480 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include <oneapi/dpl/execution>
+
+#include "support/test_config.h"
+
+#include "support/utils.h"
+
+#if _ENABLE_STD_RANGES_TESTING
+
+#include <oneapi/dpl/ranges>
+#if TEST_CPP20_SPAN_PRESENT
+#include <span>
+#endif
+#include <vector>
+#include <typeinfo>
+#include <type_traits>
+#include <string>
+
+namespace test_std_ranges
+{
+
+#if TEST_DPCPP_BACKEND_PRESENT
+template<int call_id = 0>
+auto dpcpp_policy()
+{
+    auto exec = TestUtils::default_dpcpp_policy;
+    using Policy = decltype(exec);
+    return TestUtils::make_new_policy<TestUtils::new_kernel_name<Policy, call_id>>(TestUtils::default_dpcpp_policy);
+}
+#endif //TEST_DPCPP_BACKEND_PRESENT
+
+auto host_policies() { return std::true_type{};}
+
+enum TestDataMode
+{
+    data_in,
+    data_in_out,
+    data_in_in,
+    data_in_in_out
+};
+
+auto f_mutuable = [](auto&& val) { return val *= val; };
+auto proj_mutuable = [](auto&& val) { return val *= 2; };
+
+auto f = [](auto&& val) { return val * val; };
+auto binary_f = [](auto&& val1, auto&& val2) { return val1 * val2; };
+auto proj = [](auto&& val){ return val * 2; };
+auto pred = [](auto&& val) { return val == 5; };
+auto binary_pred = [](auto&& val1, auto&& val2) { return val1 == val2; };
+
+auto pred1 = [](auto&& val) -> decltype(auto) { return val > 0; };
+auto pred2 = [](auto&& val) -> decltype(auto) { return val == 4; };
+auto pred3 = [](auto&& val) -> decltype(auto) { return val < 0; };
+
+struct P2
+{
+    P2() {}
+    P2(int v): x(v) {}
+    int x = {};
+    int y = {};
+
+    int proj() const { return x; }
+    friend bool operator==(const P2& a, const P2& b) { return a.x == b.x && a.y == b.y; }
```

Comment:
It looks like `friend` modifier is not necessary since these fields are not private.

```suggestion
    bool operator==(const P2& a, const P2& b) { return a.x == b.x && a.y == b.y; }
```

Reply:
I dont catch you intention... if you mean operator== as member functor with one argument, then:

There is a C++ rule to make user defined commutative binary operations as non-members due:

A a;
"a + 6" - valid expression
and "6 + a"  - also valid expression... but you have got an issue in case of